README.md
# MultiOgar-Edited

[![Discord](https://discordapp.com/api/guilds/407210435721560065/embed.png?style=banner2)](https://discord.gg/66X2ESb)

A slow, open source [Agar.io](http://agar.io/) server that supports multiple protocol versions and smooth vanilla physics.

Since August of 2016, [Barbosik](https://github.com/Barbosik) has stopped working on this project. So I forked the code and remade it into MultiOgar-Edited, an updated version of the previous MultiOgar.

>---

# Information
Current version : **1.6.1**

![Language](https://img.shields.io/badge/language-node.js-yellow.svg)
[![License](https://img.shields.io/badge/license-APACHE2-blue.svg)](https://github.com/Barbosik/OgarMulti/blob/master/LICENSE.md)

Original MultiOgar code is based on the private server implementation [Ogar](https://github.com/OgarProject/Ogar). The original code rightfully belongs to the [OgarProject](https://github.com/OgarProject).

MultiOgar-Edited code however, is based on MultiOgar code that has been heavily modified and improved by many collaborators. The overall goal of this fork is to make physics as vanilla as possible, cleanup most of the code, and add lots of new features while maintaining better performance than the original MultiOgar.

>---

# Installation
## Windows

1. Download and install node.js: https://nodejs.org/en/download/
2. Download this repo
3. Unzip MultiOgar-Edited code into some folder.
4. Run the win-Install_Dep.bat file.
5. Run win-Start.bat

**All these files can be found in the `run` folder.**


### Installing required modules.

```batch
:: Install Required Modules.
npm install

:: Starting the server.
cd src
node index.js
```

## Linux:
```bash
# First update your packages:
sudo apt-get update

# Install git:
sudo apt-get install git

# Install node.js:
sudo apt-get install nodejs-legacy npm

# Clone MultiOgar-Edited:
git clone git://github.com/Megabyte918/MultiOgar-Edited.git

# Install dependencies:
cd MultiOgar-Edited
npm install

# Run the server:
cd src
sudo node index.js
```

##### **For further details on running the server, please take a look inside the 'run' folder.**

>---

# Creating Issues

### Before you create your issue you ***MUST*** follow these rules:

1. **The issue is in written in English.**
2. **The issue is directly related to the project.**
3. **Your issue is not a duplicate of a previous issue.**
4. **Descriptive information must be provided, so that we can reproduce the error you are experiencing**
5. **Error logs must be provided if any crashing is experienced.**

>**Please note that we will *NOT* provide assistance if the issue is with your own code, your issue WILL be locked if this is the case.**

## Issue Formatting

**In order for us to provide you with the best assistance we kindly ask you to present your issues in the best way that you can.**

Please read through [this guide](https://guides.github.com/features/mastering-markdown) to learn how you can apply [Markdown](https://en.wikipedia.org/wiki/Markdown) to your issues.

>---

# Gallery

>### Console
>![Console](http://i.imgur.com/bS5ToRD.png)

>### Gameplay
>![Gameplay](http://i.imgur.com/XsXjT0o.png)

>---

# Performance
>Version 1.2.8 (Original MultiOgar):
>>* 1000 bots, 500 viruses, 1000 foods, map 14142x14142
>>* Works slightly slower than normal, speed decreases gradually as bots get larger.
>>* CPU load: 14% (x4 cores)
>>* Memory usage: 70 MB
>>* MS response time: Minimum of around 78

>Version 1.6.0 (MultiOgar-Edited):
>>* 1000 bots, 500 viruses, 1000 foods, map 14142x14142
>>* Works very-very smooth, speed decreases gradually as bots get larger.
>>* CPU load: 24% (x2 cores)
>>* Memory usage: 35 MB
>>* MS response time: Minimum of around 45



# Discord Bot Support

If you would like to allow a moderator to get direct access to your server's commands, you may want to set up a Discord bot on your server.

The setup is very simple, all you need is a Discord App Token and a role on your server.
Both can be specified in the config within the file.

After this is complete just run the `console-discord.js` file in the `run` folder and your server will start along with the bot.


CHANGELOG.md
# Changelog
#### 1.6.3
* Fixed safe spawning of virus, mother, and player cells

#### 1.6.2
* Optimised food and virus cell spawning and removal model (no longer use a spawning interval).
* Removed unused gamemodes.

#### 1.6.1:
* Cleaned a bit of CommandList, thanks to @AlexHGaming;
* Added botnames and removed unrealistic ones, thanks to @AlexHGaming;
* Fixed reload command;
* Added an issue template with rules, thanks to @proxiemind;
* Beautify leaderboard code and fixed board command, thanks to @Agarian;
* Added support for protocol 13, thanks to @Tyler3D;
* Added ability to separate chat for teams, thanks to @F0RIS;
* Fix minions not always collecting pellets in pellet mode;
* Added average score section to status command;
* Fixed some bugs regarding physics and collision;
* Fixed cells switching order on doublesplit, thanks to @Luka967;
* Added reading arguments on startup, thanks to @F0RIS
* Fixed skins on Cigar client, thanks to @Tyler3D;
* Improved console UI, thanks to @B0RYS and @Tyler3D;
* Tweaked the readme, thanks to @B0RYS and @MastaCoder;
* Fixed mothercell spawning issue;
* Edited the Wiki, thanks to @AlexHGaming, @Steve7914, @Tyler3D, @ItzLevvie, and @MastaCoder;
* Created the MultiOgarII Wiki, thanks to @AlexHGaming;
* Match randomskins.txt to Agar.io's skins, thanks to @ItzLevvie;

-------------
#### 1.6.0:
* Refactored collision and move engine;
* Formatted & simplified some code;
* Refactored splitting cells;
* Tweaked splitting again;
* Fixed bug with mothercells eating viruses;
* Fixed "kickbot" command kicking minions too;
* Fixed playerbots not being able to have minions;
* Reduced rate of cells switching order when doublesplitting;
* ^^ This also fixed pushback with mobile splitting;
* Completely fixed ejected mass dissapearing;
* Fixed connection issues;
* Fixed long distance popsplits;
* Got rid of vegetable bot names;
* Refactored BinaryWriter;
* Refactored BinaryReader;
* Made changes to things regarding packets, thanks to @ItzLevvie;
* Fixed food registering as playercell, thanks to @NuclearC and @ItzLevvie;
* Refactored QuadNode;
* Refactored safe spawning;
* Fixed bug with linux script;
* Fixed error with ws module;
* Made some edits to commands, thanks to @Agarian;
* Implemented vector;
* Added popsplit macro, thanks to @Tyler3D;
* Refactored some of the scripts;
* Fixed leaderboard positioning in FFA;
* Added "calc" command;
* Tweaked eject size and virus explosion multipliers;
* Fixed server not working with uws module, thanks to @Luka967

-------------
#### 1.4.8:
* Added multiple client bindings, thanks to @MastaCoder;
* Cleaned up setNickname function in Client, thanks to @MastaCoder;
* Improved popsplit/virus explosion physics yet again;
* Disabled html tags in names for console-plus, thanks to @Gimmo;
* Added mobile agar.io physics with config;
* Re-factored autosplits;
* Fixed some bugs with gamemodes;
* Simplified boosted mode calculations for cells;
* Fixed ejected mass bugs;
* Added newest agario skins to randomskins.txt;
* Created new botnames based off of names in spectate on Agar.io party;
* Added movement during linesplits, thanks to @Tyler3D;
* Fixed small bug with botplayers

-------------
#### 1.4.7:
* Added web console! All thanks to @Gimmo :)
* Cleaned more of CommandList, also thanks to @Gimmo;
* Fixed teleport command, thanks to @Agarian;
* Improved splitting physics;
* Moved changelog and license back to their original places;
* Added guide for web terminal;
* Added explode command;
* Added two new configs: clientBind and ejectVirus

-------------
#### 1.4.5:
* Added playerDecayCap config;
* Fixed some bugs;
* Refactored collision + move engine;
* Fixed cell jittering;
* Cleaned CommandList, thanks to @Gimmo;
* Improved splitting and virus explosion physics;
* Improved performance

--------------
#### 1.4.4:
* Removed verification for most player commands;
* Created the file that you are reading right now;
* Improved installation guide, cleaned up and renamed some files. Thanks to @MastaCoder;
* Greatly improved performance;
* Improved splitting and virus-explosion physics;
* Added two new commands: pop and play;
* Smartened bots, cleaned code;
* Reduced Mothercell spawn amount by 10, thanks to @Fantasy-Agario;
* Fixed some bugs

--------------
#### 1.4.2:
* Added "replace" command;
* Add srcFiles variable, thanks to @MastaCoder;
* Added ejectSpawnPercent config;
* Refactored some of the collision engine;
* Fixed annoying ejected mass bug;
* Greatly improved performance, now 2x better than original MultiOgar

--------------
#### 1.4.1:
* Fixed a few bugs;
* Added back old commands;
* Added two new commands, cleaned up "help" command;
* Slightly improved performance;
* Improved splitting and collision

--------------
#### 1.4.0:
* Added pellet mode for minions! thanks to @ZfsrGhS953;
* Cleaned/refactored the rest of GameServer;
* Improved performance;
* Added controls for pellet mode;
* Refactored minion controls;
* Created "run" folder to hold the server run files/logs;
* Added two new commands: reset and split;
* Made help command look alot better, thanks to @Agarian ;
* Skipped version 1.3.8-1.3.9 for some reason

--------------
#### 1.3.7:
* Fixed some bugs;
* Cleaned/refactored most of GameServer;
* Greatly improved performance;
* Removed/moved some files to different locations;
* Created scripts folder;
* Removed unnecessary commands aswell as functions for said commands

--------------
#### 1.3.5:
* Added new gamemode: Last Man Standing! Thanks to @Tyler3D;
* Added new minion features;
* Got rid of some more unnecessary gamemodes/commands;
* Improved performance

--------------
#### 1.3.4:
* Fixed minion bugs;
* Improved performance;
* Added random skins feature, and more

--------------
#### 1.3.3:
* Moved files;
* Fixed some bugs;
* Improved popsplits;
* Added some new features for minions

--------------
#### 1.3.0:
* Added minions! With a few new features too :)

--------------
#### 1.2.9:

src/config.js
module.exports = Object.seal({
    // MultiOgarII configurations file
// Lines starting with semicolons are comments

// [NOTES]
// MultiOgarII uses cell size instead of cell mass to improve performance!
// In order to get the cell size from mass value, you need to calculate using this formula:
//    size: SQRT( mass * 100 )
//
// For example, to set start mass: 43:
//     size: SQRT( 43 * 100 ): SQRT( 4300 ): 65.57
// Set playerStartSize: 66
//
// Also, you can use the following syntax to specify mass:
//     playerStartSize: massToSize(43)
// It will be automatically converted to 66

// [LOGGING]
// logVerbosity: Console log level (0=NONE// 1=FATAL// 2=ERROR// 3=WARN// 4=INFO// 5=DEBUG)
// logFileVerbosity: File log level
"logVerbosity": 4,
"logFileVerbosity": 5,

// [SERVER]
// serverTimeout: Seconds to keep connection alive for non-responding client
// serverWsModule: WebSocket module: 'ws' or 'uws' (install npm package before using uws)
// serverMaxConnections: Maximum number of connections to the server. (0 for no limit)
// serverPort: Server port which will be used to listen for incoming connections
// serverBind: Server network interface which will be used to listen for incoming connections (0.0.0.0 for all IPv4 interfaces)
// serverTracker: Set to 1 if you want to show your server on the tracker http://ogar.mivabe.nl/master (check that your server port is opened for external connections first!)
// serverGamemode: Gamemodes: 0: FFA, 1: Teams, 2: Experimental, 3: Rainbow
// serverBots: Number of player bots to spawn (Experimental)
// serverViewBase: Base view distance of players. Warning: high values may cause lag! Min value is 1920x1080
// serverMinScale: Minimum viewbox scale for player (low value leads to lags due to large visible area for big cell)
// serverSpectatorScale: Scale (field of view) used for free roam spectators (low value leads to lag, vanilla: 0.4, old vanilla: 0.25)
// serverStatsPort: Port for the stats server. Having a negative number will disable the stats server.
// serverStatsUpdate: Update interval of server stats in seconds
// mobilePhysics: Whether or not the server uses mobile agar.io physics
// badWordFilter: Toggle whether you enable bad word filter (set to 0 to disable)
// serverRestart: Toggle whether you want your server to auto-restart in minutes. (set to 0 to disable)
"serverTimeout": 300,
"serverWsModule": "ws",
"serverMaxConnections": 500,
"serverPort": 80,
"serverBind": "0.0.0.0",
"serverTracker": 0,
"serverGamemode": 4,
"serverBots": 0,
"serverViewBaseX": 1920,
"serverViewBaseY": 1080,
"serverMinScale": 0.15,
"serverSpectatorScale": 0.4,
"serverStatsPort": 8080,
"serverStatsUpdate": 60,
"mobilePhysics": 0,
"badWordFilter": 1,
"serverRestart": 0,

// [CLIENT]
// serverMaxLB: Controls the maximum players displayed on the leaderboard.
// serverChat: Allows the usage of server chat. 0: no chat, 1: use chat.
// serverChatAscii: Set to 1 to disable non-ANSI letters in the chat (english only)
// serverName: Server name, for example "My great server"
// serverWelcome1: First server welcome message
// serverWelcome2: Second server welcome message (optional, for info, etc)
// clientBind: Only allow connections to the server from specified client (eg: http://agar.io - http://mywebsite.com - http://more.com) [Use ' - ' to seperate different websites]
"serverMaxLB": 30,
"serverChat": 0,
"serverChatAscii": 1,
"separateChatForTeams": 0,
"serverName": "Ludecat Agar",
"serverWelcome1": "",
"serverWelcome2": "",
"clientBind": "",

// [ANTI-BOT]
// serverIpLimit: Controls the maximum number of connections from the same IP (0 for no limit)
// serverMinionIgnoreTime: minion detection disable time on server startup [seconds]
// serverMinionThreshold: max connections within serverMinionInterval time period, which will not be marked as minion
// serverMinionInterval: minion detection interval [milliseconds]
// serverScrambleLevel: Toggles scrambling of coordinates. 0: No scrambling, 1: lightweight scrambling. 2: full scrambling (also known as scramble minimap), 3 - high level scrambling (no border)
// playerBotGrow: Cells greater than 625 mass cannot grow from players under 17 mass (set to 1 to disable)
"serverIpLimit": 0,
"serverMinionIgnoreTime": 30,
"serverMinionThreshold": 10,
"serverMinionInterval": 1000,
"serverScrambleLevel": 0,
"playerBotGrow": 0,

// [BORDER]
// Border size (vanilla 14142.135623730952)
"borderWidth": 14142.135623730952,
"borderHeight": 14142.135623730952,

// [FOOD]
// foodMinSize: vanilla 10 (mass: 10*10/100: 1 mass)
// foodMaxSize: vanilla 20 (mass: 20*20/100: 4 mass)
// foodAmount: The number of food to spawn
// foodMassGrow: Enable food mass grow ?
"foodMinSize": 10,
"foodMaxSize": 20,
"foodAmount": 700,
"foodMassGrow": 1,

// [VIRUSES]
// virusMinSize: Minimum virus size. (vanilla: mass: val*val/100: 100 mass)
// virusMaxSize: Maximum virus size (vanilla: mass: val*val/100: 200 mass)
// virusMaxPoppedSize: Maximum size a popped cell can have
// virusEqualPopSize: Whether popped cells have equal size or not (1 to enable)
// virusAmount: Amount of viruses to spawn
// virusMaxAmount": Maximum amount of viruses allowed ingame.
// motherCellMaxMass: Maximum amount of mass a mothercell is allowed to have (0 for no limit)
// virusVelocity: Velocity of moving viruses (speed and distance)
// virusMaxCells: Maximum cells a player is allowed to have from virus splits (0 for playerMaxCells)
"virusMinSize": 100,
"virusMaxSize": 141.421356237,
"virusMaxPoppedSize": 60,
"virusEqualPopSize": 0,
"virusAmount": 50,
"virusMaxAmount": 100,
"motherCellMaxMass": 0,
"virusVelocity": 780,
"virusMaxCells": 0,

// [EJECTED MASS]
// ejectSize: vanilla: mass: val*val/100: 13 mass?
// ejectSizeLoss: Eject size which will be substracted from player cell (vanilla: mass: val*val/100: 18 mass?)
// ejectCooldown: Tick count until a player can eject mass again in ticks (1 tick: 40 ms)
// ejectSpawnPercent: Chance for a player to spawn from ejected mass. 0.5: 50% (set to 0 to disable)
// ejectVirus: Whether or not players can eject viruses instead of mass
// ejectVelocity: Velocity of ejecting cells (speed and distance)
"ejectSize": 36.06,
"ejectSizeLoss": 42.43,
"ejectCooldown": 3,
"ejectSpawnPercent": 0.5,
"ejectVirus": 0,
"ejectVelocity": 780,

// [PLAYERS]
// Reminder: MultiOgarII uses cell size instead of mass!
//       playerStartMass replaced with playerStartSize
//
// playerMinSize: Minimum size a player cell can decay too. (vanilla: val*val/100: 10 mass)
// playerMaxSize: Maximum size a player cell can have before auto-splitting. (vanilla: mass: val*val/100: 22500 mass)
// playerMinSplitSize: Mimimum size a player cell has to be to split. (vanilla: mass: val*val/100: 35 mass)
// playerMinEjectSize: Minimum size a player cell has to be to eject mass. (vanilla: mass: val*val/100: 35 mass)
// playerStartSize: Start size of the player cell. (vanilla: mass: val*val/100: 10 mass)
// playerMaxCells: Maximum cells a player is allowed to have. (vanilla is 16)
// playerSpeed: Player speed multiplier (1: normal speed, 2: twice the normal speed)
// playerRecombineTime: Base time in seconds before a cell is allowed to recombine (vanilla: 30 seconds)
// playerDecayRate: Amount of player cell size lost per second
// playerDecayCap: Maximum mass a cell can have before it's decayrate multiplies by 10. (0 to disable)
// playerDisconnectTime: Time in seconds before a disconnected player's cell is removed from the server (Set to -1 to never remove)
// splitVelocity: Velocity of splitting playercells (speed and distance)
"playerMinSize": 31.6227766017,
"playerMaxSize": 1500,
"playerMinSplitSize": 59.16079783,
"playerMinEjectSize": 59.16079783,
"playerStartSize": 31.6227766017,
"playerMaxCells": 16,
"playerSpeed": 1,
"playerDecayRate": 0.002,
"playerDecayCap": 0,
"playerRecombineTime": 30,
"playerMaxNickLength": 15,
"playerDisconnectTime": -1,
"splitVelocity": 780,

// [MINIONS]
// Custom minion settings
// minionStartSize: Start size of minions (mass: val*val/100: 10 mass)
// minionMaxStartSize: Maximum value of random start size for minions (set value higher than minionStartSize to enable)
// minionCollideTeam: Determines whether minions colide with their team in the Teams gamemode (0: OFF, 1: ON)
// disableERTP: Whether or not to disable ERTP controls for minions. (must use ERTPcontrol script in /scripts) (Set to 0 to enable)
// disableQ: Whether or not to disable Q controls for minions. (Set 0 to enable)
// serverMinions: Amount of minions each player gets once they spawn
// defaultName: Default name for all minions if name is not specified using command (put <r> before the name for random skins!)
// minionsOnLeaderboard: Whether or not to show minions on the leaderboard. (Set 0 to disable)
"minionStartSize": 31.6227766017,
"minionMaxStartSize": 31.6227766017,
"minionCollideTeam": 0,
"disableERTP": 1,
"disableQ": 1,
"serverMinions": 0,
"defaultName": "minion",
"minionsOnLeaderboard": 0,

// [Gamemode]
// Custom gamemode settings
// tourneyTimeLimit: Time limit of the game, in minutes.
// tourneyAutoFill: If set to a value higher than 0, the tournament match will automatically fill up with bots after value seconds
// tourneyAutoFillPlayers: The timer for filling the server with bots will not count down unless there is this amount of real players
// tourneyLeaderboardToggle Time for toggling the leaderboard, in seconds. If value set to 0, leaderboard will not toggle.
"tourneyMaxPlayers": 12,
"tourneyPrepTime": 10,
"tourneyEndTime": 30,
"tourneyTimeLimit": 20,
"tourneyAutoFill": 0,
"tourneyAutoFillPlayers": 1,
"tourneyLeaderboardToggleTime": 0

})
